File: .leetcode/46.全排列.cpp
/*
 * @lc app=leetcode.cn id=46 lang=cpp
 *
 * [46] 全排列
 */

// @lc code=start
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>>ans;
    vector<int>temp;
    void backtracking(vector<int>&nums,vector<bool>&used){
        if(temp.size()==nums.size()){
            ans.push_back(temp);
            return;
        }
        for(int i = 0;i<nums.size();i++){
            if(used[nums[i]+10]){
                temp.push_back(nums[i]);
                used[nums[i]+10] = false;
                backtracking(nums,used);
                used[nums[i]+10] = true;
                temp.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool>used(21,true);
        backtracking(nums,used);
        return ans;
    }
};
// @lc code=end

File: .leetcode/46.全排列.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> temp = new ArrayList<>();
    void dfs(int n,boolean[] visited,int[] nums){
        if(temp.size() == n){
            ans.add(new ArrayList<>(temp));
            return;
        }
        for(int i = 0;i < n;i++){
            if(!visited[i]){
                temp.add(nums[i]);
                visited[i] = true;
                dfs(n, visited, nums);
                visited[i] = false;
                temp.remove(temp.size() - 1);
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        int n = nums.length;
        boolean[] visited = new boolean[n];
        dfs(n, visited, nums);   
        return ans;
    }
}
// @lc code=end

